pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,373 @@
+
+	cnipodcache "antrea.io/antrea/pkg/agent/secondarynetwork/cnipodcache"
+	ipam "antrea.io/antrea/pkg/agent/secondarynetwork/ipam"
+	cniserver "antrea.io/antrea/pkg/agent/cniserver"
+)
+
+const (
+	controllerName = "AntreaAgentPodController"
+	minRetryDelay  = 2 * time.Second
+	maxRetryDelay  = 120 * time.Second
+	numWorkers     = 4
+)
+
+const (
+	networkAttachDefAnnotationKey = "k8s.v1.cni.cncf.io/networks"
+	cniPath                       = "/opt/cni/bin/"
+	sriov                         = "sriov"
+	netDev                        = "eth1"
+)
+
+// Set resyncPeriod to 0 to disable resyncing.
+const resyncPeriod = 0 * time.Minute
+
+type PodController struct {
+	kubeClient         clientset.Interface
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface
+	queue              workqueue.RateLimitingInterface
+	podInformer        cache.SharedIndexInformer
+	podLister          corelisters.PodLister
+	nodeName           string
+	podCache           cnipodcache.CNIPodInfoStore
+	cniServer          *cniserver.CNIServer
+}
+
+func NewPodController(kubeClient clientset.Interface,
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface,
+	podInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	nodeName string,
+	podCache cnipodcache.CNIPodInfoStore,
+	cniServer *cniserver.CNIServer) *PodController {
+
+	pc := PodController{
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pc.enqueuePod,
+			DeleteFunc: pc.enqueuePod,
+			UpdateFunc: pc.enqueuePodUpdate,
+		},
+		resyncPeriod,
+	)
+
+	pc.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podcontroller")
+	return &pc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (pc *PodController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	pc.queue.Add(podKey)
+}
+
+func (pc *PodController) enqueuePodUpdate(prevObj, newObj interface{}) {
+	return
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (pc *PodController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	pod := obj.(*corev1.Pod)
+
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
```

Comment:
I am fine if we do not support annotation change/deletion in this PR, but then could we check here if the Pod already has an secondary interface created, and if yes just log a warning message (saying we do not support change/deletion) and return here.

Reply:
Remind me when we delete secondary interfaces? Or it is to be added to this PR?

Reply:
yes, as we discussed over slack, I will raise separate PR for Annotation update/delete handle. As of now, only POD add/delete was handled.
Regarding SRIOV interface clean-up, its due by SRIOV plugin as part of the POD delete event handle. I'm discussing internally to see how we can free/clean-up SRIOV VF interface upon Annotation delete/update handle.

Reply:
Reminder for checking and ignoring Pod update events.

Reply:
yes. this was handled with a different UpdateFunc: already. Please check. (changed the handle function name accordingly).